src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/GenericDefinitionEETypeNode.cs

```diff
@@ -63,7 +54,52 @@ protected override ObjectData GetDehydratableData(NodeFactory factory, bool relo
 
             return dataBuilder.ToObjectData();
         }
+    }
+
+    internal sealed class ReflectionInvisibleGenericDefinitionEETypeNode : GenericDefinitionEETypeNode
+    {
+        public ReflectionInvisibleGenericDefinitionEETypeNode(NodeFactory factory, TypeDesc type) : base(factory, type)
+        {
+        }
+
+        public override bool ShouldSkipEmittingObjectNode(NodeFactory factory)
+        {
+            return factory.ConstructedTypeSymbol(_type).Marked;
+        }
+
+        protected override DependencyList ComputeNonRelocationBasedDependencies(NodeFactory factory)
+        {
+            return new DependencyList();
+        }
+
+        public override int ClassCode => -287423988;
+    }
+
+    internal sealed class ReflectionVisibleGenericDefinitionEETypeNode : GenericDefinitionEETypeNode
+    {
+        public ReflectionVisibleGenericDefinitionEETypeNode(NodeFactory factory, TypeDesc type) : base(factory, type)
+        {
+        }
+
+        public override bool ShouldSkipEmittingObjectNode(NodeFactory factory)
+        {
+            return false;
+        }
+
+        protected override string GetName(NodeFactory factory) => this.GetMangledName(factory.NameMangler) + " reflection visible";
+
+        protected override DependencyList ComputeNonRelocationBasedDependencies(NodeFactory factory)
+        {
+            var dependencyList = new DependencyList();
+
+            dependencyList.Add(factory.NecessaryTypeSymbol(_type), "Reflection invisible type for a visible type");
+
+            // Ask the metadata manager if we have any dependencies due to the presence of the EEType.
+            factory.MetadataManager.GetDependenciesDueToEETypePresence(ref dependencyList, factory, _type);
+
+            return dependencyList;
+        }
 
-        public override int ClassCode => -160325006;
+        public override int ClassCode => 983279111;
```

Comment:
[Q] I These are essentially random numbers, not to be duplicated across any other `GenericDefinitionEETypeNode`, but are they collected of checked for collision somewhere?

Reply:
At the end of compilation, we need to be able to sort all the outputs so that the output is stable (even though we compile on multiple threads). We know how to sort types. We know how to sort methods. But we don't know whether to sort types before methods.

These random numbers are used for that. This is what we came up with. Output needs to be stable. Comparison needs to be quick. Alternative option would be to have an enum that has all the possible classes we need to sort (there's dozens of them). We don't want a central enum with the dozens of entries that would a) become stale with orphaned entries and b) have merge conflicts all the time. These random numbers work. Accidentally making same one or forgetting to make one is a guaranteed runtime assert.

Reply:
Good to hear it's an assert!